TELCODOCS-1820: kmod changes (#74713)
openshift/openshift-docs: modules/kmm-building-and-signing-a-kmod-image.adoc

// Module included in the following assemblies:
//
// * hardware_enablement/kmm-kernel-module-management.adoc

:_mod-docs-content-type: PROCEDURE
[id="kmm-building-and-signing-a-kmod-image_{context}"]
= Building and signing a kmod image

Use this procedure if you have source code and must build your image first.

The following YAML file builds a new container image using the source code from the repository. The image produced is saved back in the registry with a temporary name, and this temporary image is then signed using the parameters in the `sign` section.

The temporary image name is based on the final image name and is set to be `<containerImage>:<tag>-<namespace>_<module name>_kmm_unsigned`.

> For example, using the following YAML file, Kernel Module Management (KMM) builds an image named `example.org/repository/minimal-driver:final-default_example-module_kmm_unsigned` containing the build with unsigned kmods and push it to the registry. Then it creates a second image named `example.org/repository/minimal-driver:final` that contains the signed kmods. It is this second image that is loaded by the `DaemonSet` object and deploys the kmods to the cluster nodes.

Reviewer: ```suggestion
For example, using the following YAML file, Kernel Module Management (KMM) builds an image named `example.org/repository/minimal-driver:final-default_example-module_kmm_unsigned` containing the build with unsigned kmods and push it to the registry. Then it creates a second image named `example.org/repository/minimal-driver:final` that contains the signed kmods. It is this second image that is pulled by the worker pods and contains the kmods to be loaded on the cluster nodes.
```

Author: Made this change.

> For example, using the following YAML file, Kernel Module Management (KMM) builds an image named `example.org/repository/minimal-driver:final-default_example-module_kmm_unsigned` containing the build with unsigned kmods and push it to the registry. Then it creates a second image named `example.org/repository/minimal-driver:final` that contains the signed kmods. It is this second image that is pulled by the worker pods and contains the kmods to be loaded on the cluster nodes.

Author: s/push/pushes

> After it is signed, the temporary image can be safely deleted from the registry. It will be rebuilt, if needed.

Author: Changed.

.Prerequisites

> * The `keySecret` and `certSecret` secrets have been created.

Reviewer 2: I would also add that the secrets needs to be created in the same namespace as the rest of the resources, to avoid confusion.

Author: Changed to "The `keySecret` and `certSecret` secrets have been created in the same namespace as the rest of the resources."

.Procedure

> . Apply the YAML file:

Author: Changed.

+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: example-module-dockerfile

>   namespace: default <1>

Author: Changed.

data:
  Dockerfile: |
    ARG DTK_AUTO
    ARG KERNEL_VERSION
    FROM ${DTK_AUTO} as builder
    WORKDIR /build/
    RUN git clone -b main --single-branch https://github.com/rh-ecosystem-edge/kernel-module-management.git
    WORKDIR kernel-module-management/ci/kmm-kmod/
    RUN make
    FROM registry.access.redhat.com/ubi9/ubi:latest
    ARG KERNEL_VERSION
    RUN yum -y install kmod && yum clean all
    RUN mkdir -p /opt/lib/modules/${KERNEL_VERSION}
    COPY --from=builder /build/kernel-module-management/ci/kmm-kmod/*.ko /opt/lib/modules/${KERNEL_VERSION}/
    RUN /usr/sbin/depmod -b /opt
---
apiVersion: kmm.sigs.x-k8s.io/v1beta1
kind: Module
metadata:
  name: example-module

>   namespace: default <1>

Author: Changed.

spec:
  moduleLoader:

>     serviceAccountName: default <2>

Author: Leaving as is. The text of the callout provides explanation.

    container:
      modprobe:
        moduleName: simple_kmod
      kernelMappings:
        - regexp: '^.*\.x86_64$'

>           containerImage: < the name of the final driver container to produce>

Reviewer: ```suggestion
          containerImage: <the name of the final driver container to produce>
```

Author: Changed.

          build:
            dockerfileConfigMap:
              name: example-module-dockerfile
          sign:
            keySecret:

>               name: <private key secret name>

Author: Changed.

            certSecret:

>               name: <certificate secret name>

Author: Changed.

            filesToSign:
              - /opt/lib/modules/4.18.0-348.2.1.el8_5.x86_64/kmm_ci_a.ko
  imageRepoSecret: <3>
    name: repo-pull-secret
  selector: # top-level selector
    kubernetes.io/arch: amd64
----

<1> `namespace` - Replace `default` with a valid namespace.

<2> `serviceAccountName` - The default `serviceAccountName` does not have the required permissions to run a module that is privileged. For information on creating a service account, see "Creating service accounts" in the "Additional resources" of this section.

> <3> `imageRepoSecret` - Used as `imagePullSecrets` in the `DaemonSet` object and to pull and push for the build and sign features.

Author: True, and it looks cleaner. I'll make these changes.
